### PR TITLE
My fork

### PR DESCRIPTION
**Pull request by VPeretiatko**

### CODE REVIEW

File: src/main/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunction.java [hunk 1]
```diff
@@ -24,19 +30,34 @@ public LinkFunction(PropertiesDao propertiesDao) {
...
+        if (propertyDto == null) {
+            LOGGER.info("Creating new Dictionary. Adding word '{}' to it.", word);
+            propertiesDao.saveGlobalProperties(new HashMap<String, String>() {{
+                put(ALTERNATIVE_DICTIONARY_PROPERTY_KEY, word);
+            }});
+        } else {
+
+            String dictionary = propertyDto.getValue();
+            ArrayList<String> wordList = new ArrayList<>(Arrays.asList(dictionary.split(SEPARATOR_CHAR)));
+            if (!wordList.contains(word)) {
```
Comment (webdizz) [inline src/main/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunction.java]: Please reverse condition, it's more natural to check for true without inversion

File: src/main/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunction.java [hunk 2]
```diff
@@ -24,19 +30,34 @@ public LinkFunction(PropertiesDao propertiesDao) {
...
+                wordList.add(word);
+                String sortedDictionary = sort(wordList);
```
Comment (webdizz) [inline src/main/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunction.java]: sort does not only sorting, please split for clearness

File: src/main/java/name/webdizz/sonar/grammar/spellcheck/GrammarChecker.java [hunk 3]
```diff
@@ -44,6 +48,9 @@ private SpellChecker createSpellChecker(final SpellCheckListener spellCheckListe
         spellCheck.addSpellCheckListener(spellCheckListener);
         spellCheck.getConfiguration().setInteger("SPELL_THRESHOLD", 1);
```
Comment (webdizz) [inline src/main/java/name/webdizz/sonar/grammar/spellcheck/GrammarChecker.java]: Let's extract this as a configuration option

File: src/main/java/name/webdizz/sonar/grammar/spellcheck/GrammarDictionaryLoader.java [hunk 4]
```diff
@@ -42,4 +45,20 @@ public SpellDictionary loadMainDictionary() {
             locker.unlock();
         }
     }
+
+    public Optional<SpellDictionaryHashMap> loadAlternateDictionary()  {
```
Comment (webdizz) [inline src/main/java/name/webdizz/sonar/grammar/spellcheck/GrammarDictionaryLoader.java]: This method shouldn't be public as we don't need it as our API

File: src/main/java/name/webdizz/sonar/grammar/spellcheck/GrammarDictionaryLoader.java [hunk 5]
```diff
@@ -42,4 +45,20 @@ public SpellDictionary loadMainDictionary() {
...
+        String alternateDict = settings.getString(PluginParameter.ALTERNATIVE_DICTIONARY_PROPERTY_KEY);
+        Reader reader;
+        if (alternateDict != null) {
+            alternateDict = alternateDict.replace(PluginParameter.SEPARATOR_CHAR, "\n");
+            reader = new StringReader(alternateDict);
+            try {
+                return Optional.of(new SpellDictionaryHashMap(reader));
+            } catch (IOException e) {
+                throw new UnableToLoadDictionary("There is no file with dictionary.", e);
```
Comment (webdizz) [inline src/main/java/name/webdizz/sonar/grammar/spellcheck/GrammarDictionaryLoader.java]: There is no file to load

File: src/main/java/name/webdizz/sonar/grammar/spellcheck/GrammarDictionaryLoader.java [hunk 6]
```diff
@@ -42,4 +45,20 @@ public SpellDictionary loadMainDictionary() {
...
+            }
+        }
+        return Optional.absent();
```
Comment (webdizz) [inline src/main/java/name/webdizz/sonar/grammar/spellcheck/GrammarDictionaryLoader.java]: There is a good practice to limit amount of returns to 1 and to not to break flow

File: src/test/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunctionTest.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+public class LinkFunctionTest {
+
+    private LinkFunction linkFunction;
+    private PropertyDto propertyDto = new PropertyDto();
+    @Mock
+    private PropertiesDao propertiesDao;
+    @Mock
+    private Function.Context context;
+    @Mock
+    Issue issue;
```
Comment (webdizz) [inline src/test/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunctionTest.java]: Make it private

File: src/test/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunctionTest.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    @Before
+    public void init() {
+        linkFunction = new LinkFunction(propertiesDao);
+        when(context.issue()).thenReturn(issue);
+        when(issue.message()).thenReturn(ERROR_DESCRIPTION + NEW_WORD + "'");
+    }
+
+    @Test
+    public void shouldCreateandUpdateNewPropertyWhenMethodInvoke() {
+        //given
```
Comment (webdizz) [inline src/test/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunctionTest.java]: It's not necessarily

File: src/test/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunctionTest.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        when(propertiesDao.selectGlobalProperty(ALTERNATIVE_DICTIONARY_PROPERTY_KEY)).thenReturn(null);
+        //when
+        linkFunction.execute(context);
+        //then
+        ArgumentCaptor<Map> property = ArgumentCaptor.forClass(Map.class);
+        verify(propertiesDao).saveGlobalProperties(property.capture());
+        assertTrue(property.getValue().containsKey(ALTERNATIVE_DICTIONARY_PROPERTY_KEY));
```
Comment (webdizz) [inline src/test/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunctionTest.java]: Too much assertions, ideally there should be 1, also use hamcrest to have more obvious test case

File: src/test/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunctionTest.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        assertEquals(NEW_WORD, property.getValue().get(ALTERNATIVE_DICTIONARY_PROPERTY_KEY));
+        assertEquals(1, property.getValue().size());
```
Comment (webdizz) [inline src/test/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunctionTest.java]: It's better to extract value with proper name

File: src/main/java/name/webdizz/sonar/grammar/spellcheck/GrammarDictionaryLoader.java (hunk 5 above)
Comment (VPeretiatko) [inline src/main/java/name/webdizz/sonar/grammar/spellcheck/GrammarDictionaryLoader.java]: we read from properties. we don't need file

File: src/test/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunctionTest.java (hunk 8 above)
Comment (VPeretiatko) [inline src/test/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunctionTest.java]: did you mean first test case? I think  it important case: *creating* property if it didn't  create before. Other cases check only updating

File: src/main/java/name/webdizz/sonar/grammar/spellcheck/GrammarDictionaryLoader.java (hunk 5 above)
Comment (webdizz) [inline src/main/java/name/webdizz/sonar/grammar/spellcheck/GrammarDictionaryLoader.java]: right, but exception tells another

File: src/test/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunctionTest.java (hunk 8 above)
Comment (webdizz) [inline src/test/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunctionTest.java]: I meant comments with give/when/then

File: src/main/java/name/webdizz/sonar/grammar/spellcheck/GrammarDictionaryLoader.java (hunk 5 above)
Comment (VPeretiatko) [inline src/main/java/name/webdizz/sonar/grammar/spellcheck/GrammarDictionaryLoader.java]: its right, but it throws by SpellDictionaryHashMap class. And now I hide it behind UnableToLoadDictionary exception

File: src/main/java/name/webdizz/sonar/grammar/spellcheck/GrammarDictionaryLoader.java [hunk 11]
```diff
@@ -42,4 +45,22 @@ public SpellDictionary loadMainDictionary() {
...
+        String alternateDict = settings.getString(PluginParameter.ALTERNATIVE_DICTIONARY_PROPERTY_KEY);
+        Reader reader;
+        Optional<SpellDictionaryHashMap> result = Optional.absent();
+        if (alternateDict != null) {
+            alternateDict = alternateDict.replace(PluginParameter.SEPARATOR_CHAR, "\n");
+            reader = new StringReader(alternateDict);
+            try {
+                result = Optional.of(new SpellDictionaryHashMap(reader));
+            } catch (IOException e) {
+                throw new UnableToLoadDictionary("There is no file with dictionary.", e);
```
Comment (webdizz) [inline src/main/java/name/webdizz/sonar/grammar/spellcheck/GrammarDictionaryLoader.java]: just rephrase an exception message to not to confuse with file as we do not have file here, for example, Cannot load dictionary from SonarQube storage

File: src/test/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunctionTest.java (hunk 8 above)
Comment (VPeretiatko) [inline src/test/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunctionTest.java]: it cost nothing, but greatly increases readable of test

File: src/main/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunction.java [hunk 12]
```diff
@@ -24,19 +29,34 @@ public LinkFunction(PropertiesDao propertiesDao) {
...
+            LOGGER.info("Creating new Dictionary. Adding word '{}' to it.", word);
+            propertiesDao.saveGlobalProperties(new HashMap<String, String>() {{
+                put(ALTERNATIVE_DICTIONARY_PROPERTY_KEY, word);
+            }});
+
+        }
+    }
+
+    private void writeSortedIfUnique(String word, String dictionary, ArrayList<String> wordList) {
+        if (!wordList.contains(word)) {
```
Comment (webdizz) [inline src/main/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunction.java]: reverse condition

File: src/main/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunction.java [hunk 13]
```diff
@@ -24,19 +29,34 @@ public LinkFunction(PropertiesDao propertiesDao) {
...
-            if (propertyDto == null){
-                LOGGER.info("Creating new Dictionary. Adding word '{}' to it.", word);
-                propertiesDao.saveGlobalProperties(new HashMap<String, String>(){{put(ALTERNATIVE_DICTIONARY_PROPERTY_KEY, word);}});
-            } else{
-                LOGGER.info("Adding word '{}' to dictionary.", word);
-                String dictionary  = propertyDto.getValue();
-                propertiesDao.updateProperties(ALTERNATIVE_DICTIONARY_PROPERTY_KEY, dictionary , dictionary + "," + word);
-            }
+        PropertyDto propertyDto = propertiesDao.selectGlobalProperty(ALTERNATIVE_DICTIONARY_PROPERTY_KEY);
+        if (propertyDto != null) {
```
Comment (webdizz) [inline src/main/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunction.java]: reverse condition

File: src/test/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunctionTest.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        when(propertiesDao.selectGlobalProperty(ALTERNATIVE_DICTIONARY_PROPERTY_KEY)).thenReturn(propertyDto);
+        //when
+        linkFunction.execute(context);
+        //then
+        verify(propertiesDao).updateProperties(ALTERNATIVE_DICTIONARY_PROPERTY_KEY, UNSORTED_WORD, NEW_WORD + SEPARATOR_CHAR + UNSORTED_WORD);
+    }
+
+    @Test
+    public void shouldNotAddWordWhenWordIsNotUnique() {
+        //given
```
Comment (webdizz) [inline src/test/java/name/webdizz/sonar/grammar/issue/tracking/LinkFunctionTest.java]: remove such comments

**Comment by webdizz** [comment]
Also please make a squash commit with good message

**Pull request merged**